reference/src/layout/scalars.md

```diff
@@ -0,0 +1,116 @@
+# Layout of scalar types
+
+This chapter represents the consensus from issue [#9]. It documents the memory
+layout and considerations for `bool`, `char`, floating point types (`f{32, 64}`), and integral types (`{i,u}{8,16,32,64,128,size}`).
+
+These types are all scalar types, representing a single value, and have no
+layout `#[repr()]` flags.
+
+[#9]: https://github.com/rust-rfcs/unsafe-code-guidelines/issues/9
+
+## `bool`
+
+Rust's `bool` has the same layout as C17's` _Bool`, that is, its size and
+alignment are implementation-defined. Any `bool` can be cast into an integer,
+taking on the values 1 (`true`) or 0 (`false`).
+
+> **Note**: on all platforms that Rust's currently supports, its size and
+> alignment are 1, and its ABI class is `INTEGER` - see [Rust Layout and ABIs].
+
+[Rust Layout and ABIs]: https://gankro.github.io/blah/rust-layouts-and-abis/#the-layoutsabis-of-builtins
+
+## `char`
+
+Rust char is 32-bit wide and represents an [unicode scalar value]. The alignment
+of `char` is _implementation-defined_.
+
+[unicode scalar value]: http://www.unicode.org/glossary/#unicode_scalar_value
+
+> **Note**: Rust `char` type is not layout compatible with C / C++ `char` types.
+> The C / C++ `char` types correspond to either Rust's `i8` or `u8` types on all
+> currently supported platforms, depending on their signedness. Rust does not
+> support C platforms in which C `char` is not 8-bit wide.
+
+## `isize` and `usize`
+
+The `isize` and `usize` types are pointer-sized signed and unsigned integers.
+They have the same layout as the [pointer types] for which the pointee is
+`Sized`, and are layout compatible with C's `uintptr_t` and `intptr_t` types.
+
+> **Note**: Rust's `usize` and C's `unsigned` types are **not** equivalent. C's
+> `unsigned` is at least as large as a short, allowed to have padding bits, etc.
+> but it is not necessarily pointer-sized.
+
+The layout of `usize` determine the following:
+
+- the maximum size of Rust values is _implementation-defined_, but can at most
+  be `usize::max_value` since `mem::size_of` and `mem::size_of_val` return
+  `usize`,
+- the maximum number of elements in an array is _implementation-defined_, but
+  can at most be `usize::max_value()` since `[T; N: usize]`,
+- the maximum value by which a pointer can be offseted is
+  _implementation-defined_, but can at most be `usize::max_value()` since
+  `ptr.add(count: usize)`.
+
+> **Note**: in the current Rust implementation:
+> 
+> * the maximum size of Rust values is limited to `isize::max_value()`. The LLVM
+> `getelementptr` instruction uses signed-integer field offsets. Rust calls
+> `getelementptr` with the `inbounds` flag which assumes that field offsets do
+> not overflow,
+> * the maximum number of elements in an array is `usize::max_value()`,
```

Comment:
It might be worth mentioning that only ZST arrays can actually be that large; non-ZST arrays would otherwise violate the first clause.